[Runtime] Create an external generic metadata builder.

mikeash opened the pull request:
Create a version of the metadata specialization code which is abstracted so that it can work in different contexts, such as building specialized metadata from dylibs on disk rather than from inside a running process.

The GenericMetadataBuilder class is templatized on a ReaderWriter. The ReaderWriter abstracts out everything that's different between in-process and external construction of this data. Instead of reading and writing pointers directly, the builder calls the ReaderWriter to resolve and write pointers. The ReaderWriter also handles symbol lookups and looking up other Swift types by name.

This is accompanied by a simple implementation of the ReaderWriter which works in-process. The abstracted calls to resolve and write pointers are implemented using standard pointer dereferencing.

A new SWIFT_DEBUG_VALIDATE_EXTERNAL_GENERIC_METADATA_BUILDER environment variable uses the in-process ReaderWriter to validate the builder by running it in parallel with the existing metadata builder code in the runtime. When enabled, the GenericMetadataBuilder is used to build a second copy of metadata built by the runtime, and the two are compared to ensure that they match. When this environment variable is not set, the new builder code is inactive.

The builder is incomplete, and this initial version only works on structs. Any unsupported type produces an error, and skips the validation.

rdar://116592420

mikeash committed:
[Runtime] Create an external generic metadata builder.

Create a version of the metadata specialization code which is abstracted so that it can work in different contexts, such as building specialized metadata from dylibs on disk rather than from inside a running process.

The GenericMetadataBuilder class is templatized on a ReaderWriter. The ReaderWriter abstracts out everything that's different between in-process and external construction of this data. Instead of reading and writing pointers directly, the builder calls the ReaderWriter to resolve and write pointers. The ReaderWriter also handles symbol lookups and looking up other Swift types by name.

This is accompanied by a simple implementation of the ReaderWriter which works in-process. The abstracted calls to resolve and write pointers are implemented using standard pointer dereferencing.

A new SWIFT_DEBUG_VALIDATE_EXTERNAL_GENERIC_METADATA_BUILDER environment variable uses the in-process ReaderWriter to validate the builder by running it in parallel with the existing metadata builder code in the runtime. When enabled, the GenericMetadataBuilder is used to build a second copy of metadata built by the runtime, and the two are compared to ensure that they match. When this environment variable is not set, the new builder code is inactive.

The builder is incomplete, and this initial version only works on structs. Any unsupported type produces an error, and skips the validation.

rdar://116592420

al45tair commented on stdlib/public/runtime/GenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,408 @@
+
+  bool isLoggingEnabled() { return true; }
+
+  __attribute((__format__(__printf__, 5, 6))) void log(const char *filename,
```
Nit: It'd be nicer to use the `[[gnu::format(printf, 5, 6)]]` syntax, I think.

al45tair commented on stdlib/public/runtime/GenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,408 @@
+              .DefaultInstantiationPattern.get());
+      auto extraDataSize = genericValueDataExtraSize(valueDescriptor, pattern);
+      if (auto *error = extraDataSize.getError()) {
+        printf("Failed to get extra data size: %s\n", error->cStr());
```
Do these want to go to `stdout`, or `stderr`, or some logging thing? And should they have some prefix to indicate where they're coming from?

mikeash replied:
I've redone this stuff to be a little more sensible. They all go to `stderr` now, and they all go through a `validationLog` function which prints a prefix on everything, with the exception of the metadata dumper which expects to be able to print partial lines and so doesn't like prefixes much.

I also decided that validation and the various logs aren't quite the same thing, so I changed it to make `SWIFT_DEBUG_VALIDATE_EXTERNAL_GENERIC_METADATA_BUILDER` an integer, where 1 does the validation silently unless there's an error, and 2 does full logging along with it.

al45tair commented on stdlib/public/runtime/GenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,408 @@
+      if (!equal) {
+        printf("Error! Mismatch between new/old metadata builders!\n");
+        printf("Original metadata:\n");
+        _swift_dumpMetadata(original);
```
(If we change these from `printf`s to anything else, we probably also need to update `_swift_dumpMetadata()` in a similar way.)

mikeash commented on include/swift/Runtime/GenericMetadataBuilder.h:
```diff
@@ -0,0 +1,965 @@
+#include <string>
+#include <variant>
+
+#define METADATA_BUILDER_LOG(...)                                              \
```
Yes it should, done.

mikeash commented on include/swift/Runtime/GenericMetadataBuilder.h:
```diff
@@ -0,0 +1,965 @@
+sizeWithAlignmentMask(uint64_t size, uint64_t alignmentMask,
+                      uint64_t hasExtraInhabitants) {
+  return (hasExtraInhabitants << 48) | (size << 16) | alignmentMask;
+}
```
`sizeWithAlignmentMask` is a weird utility we use locally to `switch` over the combination of these three values. The other two are more general, though, so I've taken them out into a new `include/swift/Basic/MathUtils.h` file.

mikeash commented on include/swift/Runtime/GenericMetadataBuilder.h:
```diff
@@ -0,0 +1,965 @@
+    else
+      errorString = "<could not create error string>";
+    free(string);
+  }
```
It does indeed complain, so I wrote it as a regular variadic function instead.

al45tair replied:
Maybe you were thinking of adding `__FILE__` and `__LINE__` to it or something.

mikeash commented:
Updated with review feedback. Also conditionalized uses of dladdr/dlsym on `SWIFT_STDLIB_HAS_DLADDR`.

mikeash commented:
@swift-ci please test

mikeash commented:
It would help if I actually added the new `MathUtils.h` file.

mikeash commented:
@swift-ci please test

al45tair commented on stdlib/public/runtime/GenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,451 @@
+  template <typename T = char>
+  BuilderErrorOr<Buffer<const T>> getSymbolPointer(const char *name) {
+#if SWIFT_STDLIB_HAS_DLADDR
+    void *ptr = dlsym(RTLD_SELF, name);
```
Linux doesn't have `RTLD_SELF`, but does have `dlsym()`, `dladdr()` and so on. They don't work as well as on Darwin, mind — they'll only find exported symbols (and they won't work at all if someone statically links with Musl, because in that situation while the APIs exist, they do nothing).

mikeash replied:
I changed this to check for `#ifdef RTLD_SELF` and use `RTLD_DEFAULT` otherwise. Only finding exported symbols is OK for our purposes. Not working at all on Musl is not, but we don't need this to work there at the moment. If we do, we replace the dlsym with a list of the functions we need to be able to find.

al45tair commented on stdlib/public/runtime/GenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,451 @@
+#include "swift/ABI/TargetLayout.h"
+#include "swift/Runtime/EnvironmentVariables.h"
+#include "swift/Runtime/Metadata.h"
+#include <dlfcn.h>
```
Should probably remove this line

```suggestion
```

mikeash replied:
Oops! Fixed.

al45tair commented on include/swift/Runtime/GenericMetadataBuilder.h:
```diff
@@ -0,0 +1,962 @@
+#include "swift/Basic/MathUtils.h"
+#include "swift/Demangling/TypeLookupError.h"
+#include "llvm/Support/Casting.h"
+#include <dlfcn.h>
```
`<dlfcn.h>` doesn't exist on Windows. We should probably conditionalise this.

```suggestion

#if SWIFT_STDLIB_HAS_DLADDR
#include <dlfcn.h>
#endif

```

al45tair commented:
The other problem looks to be some test that checks the symbols exported from `libswiftCore`.

mikeash commented:
@swift-ci please test

mikeash commented:
`SWIFT_STDLIB_HAS_DLADDR` is true when building for Windows, apparently. Rather than wrangle with the complexity of correcting that in the build system, I've added a conditional for `__has_include(<dlfcn.h>)` and we can figure out the build side separately. Windows also doesn't like the use of `vasprintf`, which is why we have our own `swift_vasprintf` which I've now switched to here.

mikeash commented:
@swift-ci please test

mikeash commented:
Windows doesn't like `flockfile`. It's not really needed here so I removed the uses.

mikeash commented:
@swift-ci please test

mikeash commented:
Need to disable the test on Windows since the lack of dlsym there causes validation to fail. We can complete Windows support separately.

mikeash commented:
@swift-ci please smoke test

mikeash commented:
Hit the `Concurrency/flow_isolation.swift` failure, test disabled in https://github.com/apple/swift/pull/70844. Retrying.

mikeash commented:
@swift-ci please smoke test
